File: synchro-mes/frontend/src/types/production.ts
// ── Production ───────────────────────────────────────────────

export type OrderStatus = 'planned' | 'in_progress' | 'completed' | 'cancelled';
export type OrderPriority = 'low' | 'normal' | 'high' | 'urgent';

export interface ProductionOrder {
  id: number;
  order_number: string;
  product_code: string;
  product_name: string;
  quantity_planned: number;
  quantity_produced: number;
  quantity_good: number;
  quantity_rejected: number;
  status: OrderStatus;
  priority: OrderPriority;
  machine_code: string | null;
  mold_code: string | null;
  operator_name: string | null;
  start_date: string | null;
  end_date: string | null;
  due_date: string | null;
  client: string | null;
  notes: string | null;
  created_at: string | null;
}

export interface ProductionOrderCreate {
  order_number: string;
  product_code: string;
  product_name: string;
  quantity_planned: number;
  priority?: string;
  machine_code?: string;
  mold_code?: string;
  due_date?: string;
  client?: string;
  notes?: string;
}

export interface ProductionOrderUpdate {
  quantity_produced?: number;
  quantity_good?: number;
  quantity_rejected?: number;
  status?: string;
  machine_code?: string;
  operator_name?: string;
  notes?: string;
}

export interface Planning {
  id: number;
  machine_code: string;
  product_code: string;
  product_name: string;
  quantity_planned: number;
  date: string;
  shift: string | null;
  mold_code: string | null;
  cycle_time_seconds: number | null;
  cavities: number;
  weight_grams: number | null;
  material: string | null;
  color: string | null;
  sequence: number;
  status: string;
  operator_name: string | null;
}

export interface PlanningCreate {
  machine_code: string;
  product_code: string;
  product_name: string;
  quantity_planned: number;
  date: string;
  shift?: string;
  mold_code?: string;
  cycle_time_seconds?: number;
  cavities?: number;
  weight_grams?: number;
  material?: string;
  color?: string;
  operator_name?: string;
  sequence?: number;
}

export interface ProductionEntry {
  id: number;
  machine_code: string;
  product_code: string;
  product_name: string | null;
  order_number: string | null;
  operator_name: string | null;
  shift: string | null;
  quantity_good: number;
  quantity_rejected: number;
  weight_kg: number | null;
  cycle_time_actual: number | null;
  timestamp: string | null;
}

export interface ProductionEntryCreate {
  machine_code: string;
  product_code: string;
  product_name?: string;
  order_number?: string;
  operator_name?: string;
  shift?: string;
  quantity_good: number;
  quantity_rejected?: number;
  weight_kg?: number;
  cycle_time_actual?: number;
  notes?: string;
}

File: synchro-mes/backend/app/schemas/production.py
"""
Schemas: ProductionOrder, Planning, ProductionEntry
"""
from pydantic import BaseModel, Field
from datetime import datetime, date
from typing import Optional


# ── Production Order ──────────────────────────────────────────
class ProductionOrderCreate(BaseModel):
    order_number: str
    product_code: str
    product_name: str
    quantity_planned: int = Field(gt=0)
    priority: str = "normal"
    machine_code: str | None = None
    mold_code: str | None = None
    due_date: date | None = None
    client: str | None = None
    notes: str | None = None


class ProductionOrderUpdate(BaseModel):
    quantity_produced: int | None = None
    quantity_good: int | None = None
    quantity_rejected: int | None = None
    status: str | None = None
    machine_code: str | None = None
    operator_name: str | None = None
    notes: str | None = None


class ProductionOrderRead(BaseModel):
    id: int
    order_number: str
    product_code: str
    product_name: str
    quantity_planned: int
    quantity_produced: int
    quantity_good: int
    quantity_rejected: int
    status: str
    priority: str
    machine_code: str | None = None
    mold_code: str | None = None
    operator_name: str | None = None
    start_date: datetime | None = None
    end_date: datetime | None = None
    due_date: date | None = None
    client: str | None = None
    notes: str | None = None
    created_at: datetime | None = None

    model_config = {"from_attributes": True}


# ── Planning ──────────────────────────────────────────────────
class PlanningCreate(BaseModel):
    machine_code: str
    product_code: str
    product_name: str
    quantity_planned: int = Field(gt=0)
    date: date
    shift: str | None = None
    mold_code: str | None = None
    cycle_time_seconds: float | None = None
    cavities: int = 1
    weight_grams: float | None = None
    material: str | None = None
    color: str | None = None
    operator_name: str | None = None
    sequence: int = 1


class PlanningRead(BaseModel):
    id: int
    machine_code: str
    product_code: str
    product_name: str
    quantity_planned: int
    date: date
    shift: str | None = None
    mold_code: str | None = None
    cycle_time_seconds: float | None = None
    cavities: int
    weight_grams: float | None = None
    material: str | None = None
    color: str | None = None
    sequence: int
    status: str
    operator_name: str | None = None

    model_config = {"from_attributes": True}


# ── Production Entry ──────────────────────────────────────────
class ProductionEntryCreate(BaseModel):
    machine_code: str
    product_code: str
    product_name: str | None = None
    order_number: str | None = None
    operator_name: str | None = None
    shift: str | None = None
    quantity_good: int = Field(ge=0)
    quantity_rejected: int = Field(ge=0, default=0)
    weight_kg: float | None = None
    cycle_time_actual: float | None = None
    notes: str | None = None


class ProductionEntryRead(BaseModel):
    id: int
    machine_code: str
    product_code: str
    product_name: str | None = None
    order_number: str | None = None
    operator_name: str | None = None
    shift: str | None = None
    quantity_good: int
    quantity_rejected: int
    weight_kg: float | None = None
    cycle_time_actual: float | None = None
    timestamp: datetime | None = None

    model_config = {"from_attributes": True}
